### PR TITLE
Unify metastore catalog interface

### PR DESCRIPTION
**Pull request by jaime0815**
issue #17041 

Co-authored-by: yun.zhang yun.zhang@zilliz.com
Co-authored-by: kejiang ke.jiang@zilliz.com

Signed-off-by: yun.zhang <yun.zhang@zilliz.com>

### CODE REVIEW

File: internal/metastore/model/collection.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package model
+
+import (
+	"github.com/milvus-io/milvus/internal/common"
+	"github.com/milvus-io/milvus/internal/proto/commonpb"
+)
+
+type Collection struct {
+	TenantID             string
```
Comment (xiaofan-luan) [inline internal/metastore/model/collection.go:9]: I was wondering what is the reason we defined model again but not use pb directly?

File: internal/rootcoord/meta_table.go [hunk 2]
```diff
@@ -1585,80 +1292,78 @@ func (mt *MetaTable) IsAlias(collectionAlias string) bool {
...
+	}
 
-	keys, _, err := mt.txn.LoadWithPrefix(CredentialPrefix)
-	if err != nil {
-		log.Error("MetaTable list all credential usernames fail", zap.Error(err))
-		return &milvuspb.ListCredUsersResponse{}, err
+	return segmentIdx, nil
+}
+
+func (mt *MetaTable) arrayConvertMap(s []common.Int64Tuple) map[int]common.Int64Tuple {
```
Comment (xiaofan-luan) [inline internal/rootcoord/meta_table.go:1355]: why this weird convert logic?

File: internal/rootcoord/meta_table.go [hunk 3]
```diff
@@ -677,126 +482,96 @@ func (mt *MetaTable) DeletePartition(collID typeutil.UniqueID, partitionName str
...
 	}
 
-	return partID, nil
+	return nil
 }
 
-// AddIndex add index; before calling this function, you must check whether indexing is required
-func (mt *MetaTable) AddIndex(segIdxInfo *pb.SegmentIndexInfo) error {
+// AlterIndex alter index
+func (mt *MetaTable) AlterIndex(newIndex *model.Index) error {
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/rootcoord/meta_table.go:554]: An index contains multiple segment indexes within the data model,  and building an index will update the segment index information, so I think it is an alter operation semantic.

File: internal/rootcoord/meta_table.go [hunk 4]
```diff
@@ -89,25 +67,30 @@ const (
...
+	indexID2Meta  map[typeutil.UniqueID]*model.Index               // collection id/index_id -> meta
 
 	ddLock   sync.RWMutex
 	credLock sync.RWMutex
 }
 
 // NewMetaTable creates meta table for rootcoord, which stores all in-memory information
 // for collection, partition, segment, index etc.
-func NewMetaTable(txn kv.TxnKV, snap kv.SnapShotKV) (*MetaTable, error) {
+func NewMetaTable(ctx context.Context, txn kv.TxnKV, snap kv.SnapShotKV) (*MetaTable, error) {
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/rootcoord/meta_table.go:88]: we will provide a factory method to choose using kv catalog or table catalog by the parameters.

File: internal/metastore/kv/constant.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package kv
+
+const (
+	// ComponentPrefix prefix for rootcoord component
+	ComponentPrefix = "root-coord"
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/metastore/kv/constant.go:5]: just for backward compatibility. although defined root-coord is wired within kv metastore, it causes cycle dependency if move it into rootcoord module.

File: internal/metastore/catalog.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	ListCollections(ctx context.Context, ts typeutil.Timestamp) (map[string]*model.Collection, error)
+	CollectionExists(ctx context.Context, collectionID typeutil.UniqueID, ts typeutil.Timestamp) bool
+	DropCollection(ctx context.Context, collectionInfo *model.Collection, ts typeutil.Timestamp) error
+
+	CreatePartition(ctx context.Context, coll *model.Collection, ts typeutil.Timestamp) error
+	DropPartition(ctx context.Context, collectionInfo *model.Collection, partitionID typeutil.UniqueID, ts typeutil.Timestamp) error
+
+	CreateIndex(ctx context.Context, col *model.Collection, index *model.Index) error
+	// AlterIndex newIndex only contains updated parts
+	AlterIndex(ctx context.Context, oldIndex *model.Index, newIndex *model.Index, alterType AlterType) error
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/metastore/catalog.go:23]: one index has multiple segment index that is described in data model

File: internal/metastore/kv/kv_catalog.go [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+	err = json.Unmarshal([]byte(v), &credentialInfo)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal credential info err:%w", err)
+	}
+
+	return &model.Credential{Username: username, EncryptedPassword: credentialInfo.EncryptedPassword}, nil
+}
+
+func (kc *Catalog) AlterAlias(ctx context.Context, collection *model.Collection, ts typeutil.Timestamp) error {
+	return kc.CreateAlias(ctx, collection, ts)
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/metastore/kv/kv_catalog.go:265]: AlterAlias is the same as CreateAlias which will update the same key

File: internal/metastore/model/collection.go (hunk 1 above)
Comment (jaime0815) [reply internal/metastore/model/collection.go:9]: [etcd Pb ](https://github.com/milvus-io/milvus/blob/015a2f08663c98b94adf2fba161d1c8c9bb4e1ba/internal/proto/etcdpb/etcd_meta.pb.go#L198) should only use within  etcd kv metastore, we need a high-level abstract model that can be used to other storage types of metastore.

File: internal/rootcoord/meta_table.go [hunk 8]
```diff
@@ -119,158 +102,82 @@ func NewMetaTable(txn kv.TxnKV, snap kv.SnapShotKV) (*MetaTable, error) {
 }
 
 func (mt *MetaTable) reloadFromKV() error {
-	mt.collID2Meta = make(map[typeutil.UniqueID]pb.CollectionInfo)
+	mt.collID2Meta = make(map[typeutil.UniqueID]model.Collection)
 	mt.collName2ID = make(map[string]typeutil.UniqueID)
 	mt.collAlias2ID = make(map[string]typeutil.UniqueID)
 	mt.partID2SegID = make(map[typeutil.UniqueID]map[typeutil.UniqueID]bool)
```
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/rootcoord/meta_table.go:108]: partID2SegID init still does not change, just keep the old implementation.  so is it wrong?

**Pull request merged**